pkg/datapath/iptables/reconciler.go

```diff
@@ -0,0 +1,203 @@
+	noTrackPods   sets.Set[noTrackPodInfo]
+}
+
+type localNodeInfo struct {
+	internalIPv4  net.IP
+	internalIPv6  net.IP
+	ipv4AllocCIDR string
+	ipv6AllocCIDR string
+}
+
+func (lni localNodeInfo) equal(other localNodeInfo) bool {
+	if !lni.internalIPv4.Equal(other.internalIPv4) ||
+		!lni.internalIPv6.Equal(other.internalIPv6) ||
+		lni.ipv4AllocCIDR != other.ipv4AllocCIDR ||
+		lni.ipv6AllocCIDR != other.ipv6AllocCIDR {
+		return false
+	}
+	return true
+}
+
+func toLocalNodeInfo(n node.LocalNode) localNodeInfo {
+	var v4AllocCIDR, v6AllocCIDR string
+
+	if n.IPv4AllocCIDR != nil {
+		v4AllocCIDR = n.IPv4AllocCIDR.String()
+	}
+	if n.IPv6AllocCIDR != nil {
+		v6AllocCIDR = n.IPv6AllocCIDR.String()
+	}
+
+	return localNodeInfo{
+		internalIPv4:  n.GetCiliumInternalIP(false),
+		internalIPv6:  n.GetCiliumInternalIP(true),
+		ipv4AllocCIDR: v4AllocCIDR,
+		ipv6AllocCIDR: v6AllocCIDR,
+	}
+}
+
+type proxyInfo struct {
+	name        string
+	port        uint16
+	isLocalOnly bool
+}
+
+type noTrackPodInfo struct {
+	ip   netip.Addr
+	port uint16
+}
+
+func reconciliationLoop(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	health cell.HealthReporter,
+	installIptRules bool,
+	params *reconcilerParams,
+	updateRules func(state desiredState, firstInit bool) error,
+) error {
+	// The minimum interval between reconciliation attempts
+	const minReconciliationInterval = 200 * time.Millisecond
+
+	state := desiredState{
+		installRules: installIptRules,
+		proxies:      sets.New[proxyInfo](),
+		noTrackPods:  sets.New[noTrackPodInfo](),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	localNodeEvents := stream.ToChannel(ctx, params.localNodeStore)
+	state.localNodeInfo = toLocalNodeInfo(<-localNodeEvents)
+
+	devices, devicesWatch := tables.SelectedDevices(params.devices, params.db.ReadTxn())
+	state.devices = sets.New(tables.DeviceNames(devices)...)
+
+	// Use a ticker to limit how often the desired state is reconciled to avoid doing
+	// lots of operations when e.g. ipset updates.
+	ticker := time.NewTicker(minReconciliationInterval)
+	defer ticker.Stop()
+
+	// stateChanged is true when the desired state has changed or when reconciling it
+	// has failed. It's set to false when reconciling succeeds.
+	stateChanged := true
+
+	firstInit := true
+
+	if err := updateRules(state, firstInit); err != nil {
+		health.Degraded("iptables rules update failed", err)
+		// Keep stateChanged=true and firstInit=true to try again on the next tick.
+	} else {
+		health.OK("iptables rules update completed")
+		firstInit = false
+		stateChanged = false
+	}
+
+stop:
+	for {
+		select {
+		case <-ctx.Done():
+			break stop
+		case <-devicesWatch:
+			devices, devicesWatch = tables.SelectedDevices(params.devices, params.db.ReadTxn())
+			newDevices := sets.New(tables.DeviceNames(devices)...)
+			if newDevices.Equal(state.devices) {
+				continue
+			}
+			state.devices = newDevices
+			stateChanged = true
+		case localNode, ok := <-localNodeEvents:
```

Comment:
`localNodeEvents` will close when the reconciler `ctx` closes. That `ctx` is the one from the job, so it will be closed when the job group stop hook is called. All the other channels `params.proxies`, `params.addNoTrackPod` and `params.delNoTrackPod` are closed when the `manager.Stop()` is called, that is, when the hive is stopped.
IOW, we can break out of the loop for that reason only.